### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by thebladehit**

### CODE REVIEW

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  ) {}
+
+  async register (dto: RegisterDto) {
+    const user = await this.usersRepository.findByEmail(dto.email);
+    if (user) {
+      throw new ConflictException('A user with this email already exists');
+    }
+    delete dto.passwordRepeat;
+    const hashedPassword = await this.hashPassword(dto.password);
+    dto.password = hashedPassword;
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:31]: Local variable `hashedPassword` is redundant => you can get rid of it

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    const token = uuidv4();
+    const expireDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
+    const verifiedToken = `${token}::email::${expireDate}`;
+    dto['verifiedToken'] = verifiedToken;
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:35]: Local variable `verifiedToken` is redundant  => you can get rid of it

File: src/auth/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    const url = process.env.FRONTEND_DOMAIN + 'verify-email/' + token;
+    await this.mailService.sendVerifyEmail(dto.email, url);
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:37]: Your `sendVerifyEmail` is not a async function => get rid of redundant `await`

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      throw new NotFoundException(`No user found for email: ${dto.email}`);
+    }
+    if (user.verifiedToken !== null && user.verifiedToken.includes('::email::')) {
+      throw new UnauthorizedException('Verify email');
+    }
+    const token = uuidv4();
+    const expireDate = new Date(Date.now() + 15 * 60 * 1000);
+    const verifiedToken = `${token}::password::${expireDate}`;
+    const url = process.env.FRONTEND_DOMAIN + 'reset-password/' + token;
+    await this.mailService.sendResetPassword(dto.email, url);
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:109]: Your `sendResetPassword` is not an async func => get rid of redundant `await`

File: src/auth/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    if (!user) {
+      throw new BadRequestException(`Invalid token: ${dto.token}`);
+    }
+    const verifiedTokenSplited = user.verifiedToken.split('::');
+    const expirationDate = verifiedTokenSplited[verifiedTokenSplited.length - 1];
+    if (new Date() > new Date(expirationDate)) {
+      throw new BadRequestException('Password verification token is expired');
+    }
+    const hashedPassword = await this.hashPassword(dto.password);
+    dto.password = hashedPassword;
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:124]: Local variable `hashedPassword` is redundant => you can get rid of it

File: src/auth/auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    const isPasswordsValid = await bcrypt.compare(dto.password, user.password);
+    if (!isPasswordsValid) {
+      throw new UnauthorizedException('Invalid password');
+    }
+    const token = uuidv4();
+    const expireDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
+    const verifiedToken = `${token}::email::${expireDate}`;
+    const url = process.env.FRONTEND_DOMAIN + 'verify-email/' + token;
+    await this.mailService.sendVerifyEmail(user.email, url);
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:144]: Your `sendVerifyEmail` is not a async function => get rid of redundant `await`

File: src/auth/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        userId,
+        hash: this.getHash(email, password),
+      }),
+      refreshToken: uuidv4(),
+    };
+  }
+
+  getHash (...params) {
+    return crypto.createHash('md5').update(params.join()).digest('hex');
+  }
```
Comment (SharpDevOps10) [inline src/auth/auth.service.ts:160]: add type for `getHash`

```ts
(...params: string[]) 
```

**Review by SharpDevOps10 — CHANGES_REQUESTED** [review]

**Review by SharpDevOps10 — APPROVED** [review]

**Review by KobbAsa — COMMENTED** [review]
Not bad.

**Review by KobbAsa — APPROVED** [review]

**Pull request merged**